### ISSUE TITLE
[Icons]: A modest proposal: Add icons to the standards

### ISSUE THREAD

**Issue by esgoodman** [uswds/uswds#762]
Many federal websites have similar features and sections. Yet the icons and treatments vary wildly. As someone who makes lots of quick comps, it would streamline my work enormously to have a set of government-directed icons available that are free to use and cover some of those basic functions. 

18F/OCSIT has created lots and lots of badges and icons in the short time we've existed. Not to mention the rest of the federal government. What I imagine would make sense would be:
- start collecting icons from federal websites
- cluster into common concepts
- select the icon metaphor that best communicates the concept, or draw a new one
- standardize the icons to match the rest of the standards.

**Comment by anthonygarvan** [comment]
+1

**Comment by scoota69** [comment]
Also fonts would be appreciated along with the icons

**Comment by mollieru** [comment]
We had been leaning toward font-awesome for icons @esgoodman -- any thoughts on this?  Thread on icons here: https://github.com/18F/web-design-standards/issues/43

**Comment by KimberlyMunoz** [comment]
It doesn't encompass everything, but CFPB has open-sourced icons with an icon font at https://github.com/cfpb/cf-icons.

**Comment by MsWebGirl** [comment]
+1 @KimberlyMunoz as a agency that presents a lot of financial information, having specific govt-wide financial icons for instance, would be helpful

**Comment by esgoodman** [comment]
That's exactly what I was thinking. FontAwesome is a generalist set of icons that works great for general interest sites. It doesn't present a wide range of options for many subjects that concern government websites, such as financial transactions, sign-in/identity management (my own concern), security, etc.

**Comment by joshbruce** [comment]
For the program I'm working with, we do include FontAwesome and ask our designers to follow the style of those icons. Therefore, the FontAwesome icons won't clash with our custom work - and we don't have to design icons already covered by FontAwesome.

**Comment by JHKennedy4** [comment]
+1 from Vote.org

We send our <3.

**Comment by mattlangan** [comment]
What are the accessibility implications (thinking screen readers here) of adding the <i> tags where the icons are used? It's my understanding that I should add aria-hidden="true" to all such tags, which could get a bit cumbersome. I'll likely begin using a helper method to render all such tags but wanted to explicitly state that point of concern.

FWIW I've had great experience working with font-awesome and just integrated it seamlessly into my latest USWDS-based project.

**Comment by joshbruce** [comment]
@mattlangan - Your mentioned of "helper methods" brings me to ask: What language(s) are you using on your project?

The underlying destination is that the program I am on has been working on kit containing configurable helpers to return USWDS-based components. Think template engine - only independent of framework - and for components, not whole pages.

**Comment by mattlangan** [comment]
@joshbruce I'm using Ruby on Rails. The majority of places where I'm rendering an icon are already using helper methods (nav links) to run additional logic like applying .usa-current. I can just add an additional param to the request to pass the icon declaration. I think you and I are thinking along similar lines with regard to _generating_ html without _writing_ it :)

I've also been tinkering with what I'm calling "Protosketching on Rails", which is to be a pre-configured Rails scaffold that includes core features like USWDS, Devise w/ User model, etc... It's more of a focus on feature scaffolding rather than UI, but still leveraging the underlying UI framework that comes out of the USWDS box.

**Comment by joshbruce** [comment]
@mattlangan - Nice! And agreed. 

My program is still figuring out if we are going to change our server-side framework. As a proof-of-concept, this is what we put together:

https://github.com/joshbruce/iae-uikit-php/blob/master/src/FormControls/Select.php

I've iterated a lot on the concepts since then, but the basics are still in play. If you're not familiar with PHP the short version is - you feed it a configuration key-value array (Hash or even JSON). It returns the  HTML - similar to they UI Kit works for iOS development.

If we decide to go with Rails - would you be interested in something like that? 

ps. I'm trying to gauge demand outside of my program. If demand is low, we would make one gem (in the case of Ruby) specific to us, and still open to the public. If demand is moderate to high though, we might split it into two gems - one generic to USWDS components and the program-specific one that is dependent on the first, again, open to the public.

**Comment by Drakenhart** [comment]
https://dusted.codes/making-font-awesome-awesome-using-icons-without-i-tags

Something to consider?

**Comment by mattlangan** [comment]
@joshbruce will move convo to PM to unhijack this thread :)

@Drakenhart Yes that definitely looks compelling, especially since I've heard that the aria hidden declaration might not be 100% reliable. I also believe that CSS-generated content is ignored by screen readers, which is precisely what we would want in this case.

From http://www.ssbbartgroup.com/blog/csscontentproperty/
> Now those of you who know a little bit about screen readers are probably having a moment of truth right now and/or a conniption. This is because screen reader software usually reads from the DOM that initially loads in the browser before the stylesheet is applied. The symbols that are inserted into the page using the content CSS property are generated symbols loaded after the DOM loads.  The screen reader will not announce generated content to the user who is blind — even if the user refreshes the screen reader’s buffer after the page loads. This can be a problem if the symbols inserted into the page are essential to understand the meaning of the page. If the content CSS property is used to insert a little smiley face, for example, it’s no big loss to the screen reader user if he doesn’t know it’s there.

I'd want to dig in to make sure that change isn't at the expense of using the <i> tag (which might still have a place... or might not) but the approach you linked certainly seems to eliminate excess tagging without changing the way FA is working and without running into font-family conflict.

**Comment by Drakenhart** [comment]
@mattlangan Also it's good semantics as the "i" tag _(wasn't previously used but)_ seems to have been re-worked.

https://www.w3.org/International/questions/qa-b-and-i-tags
https://www.w3.org/TR/html5/text-level-semantics.html#the-i-element

**Comment by mattlangan** [comment]
@Drakenhart well that ends the discussion for me. gonna re-work to strip <i> tags. thank you very much!

**Comment by joshbruce** [comment]
@mattlangan - I didn't think there was a PM capability here, but I agree on the hijacking. I've proposed something similar to a few folks in the USWDS community, and have felt bad every time. So, I made the following for it:

https://github.com/joshbruce/iae-uikit-php/issues/5

**Comment by ostranme** [comment]
Another shameless plug for http://govicons.io/. Might be worth considering as an option.

**Comment by joshbruce** [comment]
Funny story as I completely shift from my original position in #43. I've actually removed all icons and gone to text.

A few reasons, if I may:

1. They simple don't really add much to the UI and experience, at least not for my purposes. (I do use a text editor plug-in that has icons, but that's not my UI, technically.)
2. I was using all of two (this is for the platform I'm developing than the one discussed in #43). An icon, when done well (and totally in my opinion), communicates it's purpose, and there just weren't a lot of areas where the icon-set could actually do that nor could I figure out an icon to create that would.
3. Given I was using two icons and bringing in the netire Font Awesome base CSS, seems a bit much.

So, if we go with icons and all that, I think it might be worth putting in the guidance regarding usage something like: If you have room for text, use text. Of course, the con here is in localization and internationalization. :)

Also, just found this while looking for another WWDC session (on the difference between glyph and icon and other design tasties): 

https://developer.apple.com/videos/play/wwdc2017/823/

**Comment by s2t2** [comment]
As a developer using USWDS for the first time, I was kind of expecting Font Awesome to be in here. It seems FA icons are used in [uswds-assets](https://github.com/uswds/uswds-assets), but not in the standard itself.

The argument for inclusion is reasonable. Font Awesome is perhaps the most popular set of icons on the web. And Twitter Bootstrap has included an icon font in their framework for years. Inclusion provides developers with a **consistent and effort-minimizing out-of-the-box experience**.

However the argument against inclusion is also reasonable. FA is not the only possible icon set and is not necessarily comprehensive for government-purposes (as discussed in comments above). Also, Twitter Bootstrap has [removed icon fonts](https://v4-alpha.getbootstrap.com/migration/#components) from their newest version (4.0).

A good compromise would be to at least add information to the USWDS docs providing links to various icon sets, and guidance on the recommended approach for how to include and use them in conjunction with USWDS.

**Comment by maya** [comment]
Agree with @s2t2 that a good lightweight approach would be an icons page describing that we use Font Awesome and how to add them to your project. We include a subset of FA icons in the image assets folder that the design system uses.

**Comment by maya** [comment]
Thanks all! We went with the lightweight approach which you can now view here: https://designsystem.digital.gov/components/icons/
